src/lib.rs

```diff
@@ -8,6 +8,244 @@ pub mod qdrant;
+
+use std::error::Error;
+use std::fmt::{Debug, Display, Formatter};
+
+static NULL_VALUE: Value = Value {
+    kind: Some(NullValue(0)),
+};
+
+macro_rules! get_payload {
+    ($ty:ty) => {
+        impl $ty {
+            /// get a payload value for the specified key. If the key is not present,
+            /// this will return a null value
+            pub fn get(&self, key: &str) -> &Value {
+                self.payload.get(key).unwrap_or(&NULL_VALUE)
+            }
+        }
+    };
+}
+
+get_payload!(RetrievedPoint);
+get_payload!(ScoredPoint);
+
+macro_rules! extract {
+    ($kind:ident, $check:ident) => {
+        /// check if this value is a
+        #[doc = stringify!($kind)]
+        pub fn $check(&self) -> bool {
+            matches!(self.kind, Some($kind(_)))
+        }
+    };
+    ($kind:ident, $check:ident, $extract:ident, $ty:ty) => {
+        extract!($kind, $check);
+
+        /// extract the contents if this value is a
+        #[doc = stringify!($kind)]
+        pub fn $extract(&self) -> Option<$ty> {
+            if let Some($kind(v)) = self.kind {
+                Some(v)
+            } else {
+                None
+            }
+        }
+    };
+    ($kind:ident, $check:ident, $extract:ident, ref $ty:ty) => {
+        extract!($kind, $check);
+
+        /// extract the contents if this value is a
+        #[doc = stringify!($kind)]
+        pub fn $extract(&self) -> Option<&$ty> {
+            if let Some($kind(v)) = &self.kind {
+                Some(v)
+            } else {
+                None
+            }
+        }
+    };
+}
+
+impl Value {
+    extract!(NullValue, is_null);
+    extract!(BoolValue, is_bool, as_bool, bool);
+    extract!(IntegerValue, is_integer, as_integer, i64);
+    extract!(DoubleValue, is_double, as_double, f64);
+    extract!(StringValue, is_str, as_str, ref String);
+    extract!(ListValue, is_list, as_list, ref [Value]);
+    extract!(StructValue, is_struct, as_struct, ref Struct);
+
+    #[cfg(feature = "serde")]
+    /// convert this into a `serde_json::Value`
+    pub fn into_json(self) -> serde_json::Value {
+        use serde_json::Value as JsonValue;
+        match self.kind {
+            Some(BoolValue(b)) => JsonValue::Bool(b),
+            Some(IntegerValue(i)) => JsonValue::from(i),
+            Some(DoubleValue(d)) => JsonValue::from(d),
+            Some(StringValue(s)) => JsonValue::String(s),
+            Some(ListValue(vs)) => vs.into_iter().map(Value::into_json).collect(),
+            Some(StructValue(s)) => s
+                .fields
+                .into_iter()
+                .map(|(k, v)| (k, v.into_json()))
+                .collect(),
+            Some(NullValue(_)) | None => JsonValue::Null,
+        }
+    }
+}
+
+#[cfg(feature = "serde")]
+impl From<Value> for serde_json::Value {
+    fn from(value: Value) -> Self {
+        value.into_json()
+    }
+}
+
+impl Display for Value {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
```

Comment:
Same here, a simple unit test to validate that it does what it says would be great.